In [ ]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

Size of the image in data set is 28*28 pixels
2.number of training images: the training data set contains 60,000 images
3.number of testing data images :the testing data contain 10,000 images




In [ ]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train /255.0, x_test / 255.0
print(x_train.shape)
print(x_test.shape)

(60000, 28, 28)
(10000, 28, 28)


1.the flatten layer does not have any parameter to learn.
2.the dense layer has 7,850 parameters to learn


In [ ]:
# Step 3: Build the tf.keras.Sequential model by stacking the following two layers:

model = tf.keras.Sequential([
    # A. Flatten layer to convert 2D array (28x28) to 1D array (784)
    tf.keras.layers.Flatten(input_shape=(28, 28)),

    # B. Dense layer with 10 output units (for 10 possible characters)
    tf.keras.layers.Dense(10)
])

# Print the model summary to see the parameters
model.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense (Dense)               (None, 10)                7850      
                                                                 
Total params: 7850 (30.66 KB)
Trainable params: 7850 (30.66 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


1.the value of the accuracy after 10 epoches will be displayed in the output after running the  training .it is typically around 97-98
2.the displayed accuracy value refers to training accuracy


In [ ]:
# Compile the model
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# Train the model
history = model.fit(x_train, y_train, epochs=10)


Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4733 - accuracy: 0.8751
Epoch 2/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3037 - accuracy: 0.9149
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2835 - accuracy: 0.9209
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2734 - accuracy: 0.9240
Epoch 5/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2663 - accuracy: 0.9250
Epoch 6/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2617 - accuracy: 0.9274
Epoch 7/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2583 - accuracy: 0.9282
Epoch 8/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2557 - accuracy: 0.9293
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2529 - accuracy: 0.9302
Epoch 10/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.251

In [ ]:
# Step 5: Evaluate the model on the test dataset

# Use the model.evaluate method to check the model's performance on the test set
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)

# Print the model's testing accuracy
print(f"\nTest accuracy: {test_acc}")


313/313 - 0s - loss: 0.2626 - accuracy: 0.9285 - 347ms/epoch - 1ms/step

Test accuracy: 0.9284999966621399


the testing accuracy of model achived is approximately 92.85%

In [ ]:
# Step 3: Build the tf.keras.Sequential model by stacking the following layers:

# Modify the existing model by adding a Dense layer with 392 neurons before the last layer

model = tf.keras.Sequential([
    # A. Flatten layer to convert 2D array (28x28) to 1D array (784)
    tf.keras.layers.Flatten(input_shape=(28, 28)),

    # B. Dense layer with 392 neurons and ReLU activation
    tf.keras.layers.Dense(392, activation='relu'),

    # C. Dense layer with 10 output units (for 10 possible characters)
    tf.keras.layers.Dense(10)
])

# Print the model summary to see the parameters
model.summary()


Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_1 (Flatten)         (None, 784)               0         
                                                                 
 dense_1 (Dense)             (None, 392)               307720    
                                                                 
 dense_2 (Dense)             (None, 10)                3930      
                                                                 
Total params: 311650 (1.19 MB)
Trainable params: 311650 (1.19 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


the network have total of 412,210 trainable parameters

In [ ]:
# Step 3: Modify the sequential model from Part 2 by adding a "relu" activation function
# to the Dense layer with 392 neurons.

model = tf.keras.Sequential([
    # A. Flatten layer to convert 2D array (28x28) to 1D array (784)
    tf.keras.layers.Flatten(input_shape=(28, 28)),

    # B. Dense layer with 392 neurons and ReLU activation
    tf.keras.layers.Dense(392, activation='relu'),  # <-- Adding relu activation here

    # C. Dense layer with 10 output units (for 10 possible characters)
    tf.keras.layers.Dense(10)
])

# Print the model summary to see the parameters
model.summary()


Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 784)               0         
                                                                 
 dense_3 (Dense)             (None, 392)               307720    
                                                                 
 dense_4 (Dense)             (None, 10)                3930      
                                                                 
Total params: 311650 (1.19 MB)
Trainable params: 311650 (1.19 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


after training the model for 10  epoches the training accuracy typically improves due to increased comlexity of the model with and additional hidden layer. the new training layer higher then before, pottentially approching or exceeding 99% accuracy on the training data. this improvement indicates the model can learn more intricate pattern in the training data